<a href="https://colab.research.google.com/github/Anshuman776/Deep-Learning-Projects-/blob/main/ANN_Implementation_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [63]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [64]:
dataset = pd.read_csv("/content/Churn_Modelling.csv")
x = dataset.iloc[:,3:13].values
y = dataset.iloc[:,13].values

In [65]:
dataset.head(5)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender   Age  \
0          1    15634602  Hargrave          619    France  Female  42.0   
1          2    15647311      Hill          608     Spain  Female  41.0   
2          3    15619304      Onio          502    France  Female  42.0   
3          4    15701354      Boni          699    France  Female  39.0   
4          5    15737888  Mitchell          850     Spain  Female  43.0   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [66]:
x

array([[619, 'France', 'Female', ..., 1, 1, 101348.88],
       [608, 'Spain', 'Female', ..., 0, 1, 112542.58],
       [502, 'France', 'Female', ..., 1, 0, 113931.57],
       ...,
       [709, 'France', 'Female', ..., 0, 1, 42085.58],
       [772, 'Germany', 'Male', ..., 1, 0, 92888.52],
       [792, 'France', 'Female', ..., 1, 0, 38190.78]], dtype=object)

In [67]:
y

array([1, 0, 1, ..., 1, 1, 0])

In [68]:
# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.impute import SimpleImputer # Corrected import path
import numpy as np

labelencoder_x_1 = LabelEncoder()
x[:,1] = labelencoder_x_1.fit_transform(x[:,1])
labelencoder_x_2 = LabelEncoder()
x[:,2] = labelencoder_x_2.fit_transform(x[:,2])

# Explicitly convert the array to a numeric type (float) after all encoding
x = x.astype(float)

# Handle any remaining NaN values in x using SimpleImputer before splitting/scaling
imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
x = imputer.fit_transform(x)

x

array([[6.1900000e+02, 0.0000000e+00, 0.0000000e+00, ..., 1.0000000e+00,
        1.0000000e+00, 1.0134888e+05],
       [6.0800000e+02, 2.0000000e+00, 0.0000000e+00, ..., 0.0000000e+00,
        1.0000000e+00, 1.1254258e+05],
       [5.0200000e+02, 0.0000000e+00, 0.0000000e+00, ..., 1.0000000e+00,
        0.0000000e+00, 1.1393157e+05],
       ...,
       [7.0900000e+02, 0.0000000e+00, 0.0000000e+00, ..., 0.0000000e+00,
        1.0000000e+00, 4.2085580e+04],
       [7.7200000e+02, 1.0000000e+00, 1.0000000e+00, ..., 1.0000000e+00,
        0.0000000e+00, 9.2888520e+04],
       [7.9200000e+02, 0.0000000e+00, 0.0000000e+00, ..., 1.0000000e+00,
        0.0000000e+00, 3.8190780e+04]])

In [69]:
# spliting the dataset info the training set and test set
from sklearn.model_selection import train_test_split
x_train , x_test , y_train , y_test = train_test_split(x,y,test_size = 0.2, random_state=0)

In [70]:
len(dataset)

10000

In [71]:
len(x_test)

2000

In [72]:
# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test = sc.transform(x_test)

In [73]:
# classical Ml
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier()
clf.fit(x_train,y_train)
pred = clf.predict(x_test)


In [74]:
#  making confusion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test ,pred)
print(cm)

[[1360  235]
 [ 193  212]]


In [75]:
#  making the confusion matrix
from sklearn.metrics import accuracy_score
score = accuracy_score(y_test,pred)
print(score*100) # this 100 to show percentage

78.60000000000001


Accuracy from Decision TREE classifier


In [76]:
# Random forest
# classical Ml

from sklearn.ensemble import RandomForestClassifier
clf2 = RandomForestClassifier(n_estimators=200)
clf2.fit(x_train,y_train)
pred2 = clf2.predict(x_test)


In [77]:
# making the confusion matrix
from sklearn.metrics import confusion_matrix
cm2 = confusion_matrix(y_test , pred2)
print(cm2)

[[1529   66]
 [ 203  202]]


In [78]:
# checking accuracy
from sklearn.metrics import accuracy_score
score2 = accuracy_score(y_test,pred2)
print(score2*100)

86.55000000000001


**Accuracy of Random Forest classifier is 86%**

**ANN**

In [79]:
# importing the keras libraries and packages
import keras
from keras.models import Sequential
from keras.layers import Dense

In [80]:
# Initialising the ANN
import keras
from keras.models import Sequential
from keras.layers import Dense, Input

classifier = Sequential()
# Adding the input layer and the first hidden layer
classifier.add(Input(shape=(10,))) # Explicitly defining the input layer
classifier.add(Dense(units=6,kernel_initializer ='uniform',activation='relu'))
# Adding the second hidden layer
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
# Adding the third layer
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
# Adding the output layer
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

In [81]:
len(x_train)

8000

In [82]:
# compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train[0]
# fitting the ANN to the training set
classifier.fit(x_train,y_train,batch_size=10,epochs=50)

Epoch 1/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7899 - loss: 0.5630
Epoch 2/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7894 - loss: 0.4411
Epoch 3/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7919 - loss: 0.4249
Epoch 4/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8065 - loss: 0.4187
Epoch 5/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8309 - loss: 0.4095
Epoch 6/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8252 - loss: 0.4235
Epoch 7/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8260 - loss: 0.4169
Epoch 8/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8231 - loss: 0.4305
Epoch 9/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8272 - loss: 0.4285
Epoch 10/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8351 - loss: 0.4172
Epoch 11/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8351 - loss: 0.4129
Epoch 12/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step

In [83]:
# Making prediction and evaluating the model
# predicting the test set result
y_pred_ann = classifier.predict(x_test)
y_pred_ann = (y_pred_ann > 0.5)

cm3 = confusion_matrix(y_test,y_pred_ann)
print(cm3)


63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step
[[1544   51]
 [ 271  134]]


In [84]:
from sklearn.metrics import accuracy_score
score3 = accuracy_score(y_test,y_pred_ann)
print(score3*100)

83.89999999999999


**Accuracy from ANN:84%**

**Conclusion**

so,as we have alredy seen , by using 3 hidden layered ANN, we aren't able to get beyond 84% accuracy , ofcourse we can get higher accuracy by increasing the epochs & doing some hyperparameter tuning , so on the other hand , the classical models can also be tuned futher .

Bottomline, always start with classical machine learning models, and if we aren't able to increse the performance , then move towards Deep Learning techniques.